### DISCUSSION TITLE
Support xz, bz2 and lz4 compressed CSV files

### DISCUSSION THREAD

**Discussion by ivbeg** [duckdb/duckdb#12232] [Ideas]
Right now DuckDB supports Gzipped or uncompressed CSV files only, but too often CSV files stored with other compressions like bzip2, xz or lz4. Please support other compressions too.

**Comment by EvanZ** [comment]
I have the same issue needing bzip2 support with json or csv. We use this compression quite often in our Spark flows.

**Comment by rbpasker** [comment]
consider using Arrow to read the CSVs https://arrow.apache.org/docs/python/generated/pyarrow.Codec.html#pyarrow.Codec

**Comment by rustyconover** [comment]
Actually with the shellfs extension at:

https://community-extensions.duckdb.org/extensions/shellfs.html

You can put any decompression program before read_csv() that you want as long as it writes to standard out.

I wrote the extension that lets this happen, happy to answer any questions.

**Reply by michael-doubez** [reply]
That supposes a shell with bunzip is available

**Reply by ivbeg** [reply]
Thanks, that's useful.
It require additional control over execution environment. Since analytics use Linux, Mac OS X and Windows, and execution environment could include any Unix-style server, it's really useful only for a certain cases, not universal(

**Comment by michael-doubez** [comment]
There almost everything needed for compression extension: a registration system in virtual filesystem and an interface.
It is the interaction with the filesystem which is blocking:

- compression id is based on enum; uint8 which is far too small to have generic ids
- file extension identification is hard coded

A simple hack to minimize the break or interface would be to increase the compression id to uint32 (or even a uint16):
- id 0 to 32 reserved for core compression (FileCompressionType::*)
- remainder identify compression by taking the extension name as integral (2+ ascii casted into integral)
  - bz2 --> 0x62,0x7a,0x32,0x00
  - xz --> 0x78,0x7a,0x00,0x00
  - lz(4) --> 0x6C,0x7a,0x34,0x00
 
In AUTO mode, we can simply look at the extension of the file and /cast/ it to integral for a lookup (some compressions could be registered multiple times if multiple extension apply).

**Comment by BasixKOR** [comment]
As this request is quite similar to #14842 and it's likely this would be brought up for any compression algorithm, could we consider adopting [libarchive](https://github.com/libarchive/libarchive) instead? It would solve a lot of issues regarding compressions.

**Comment by dentiny** [comment]
I think duckdb discourages external dependency in core (see https://duckdb.org/2024/03/22/dependency-management.html),
and as of now it's tightly coupled with filesystem, see https://github.com/duckdb/duckdb/blob/main/src/common/gzip_file_system.cpp
so the recommended way I guess might be adding new extensions for new compression algorithms and integrate with VFS.

**Comment by ivbeg** [comment]
Well, I have some popularity statistics for the most common file extensions from our data search engine Dateno. Here https://github.com/datenoio/dateno-stats/blob/main/data/current/stats_formats.csv 

From open data catalogues content view most popular compression file formats are zip, gzip, 7zip, xz, bz2 and there are also several compression algorithms popular among data engineers: lz4, brotli, snappy, lzo. From my personal experience, xz, bz2 and lz4 are most commonly used after gzip and zip.

**Comment by nickzoic** [comment]
Would it be impossible to add bzip2 to the third_party libraries to support this?  The rest of the work required doesn't look all that difficult if I just follow @Tishj 's lead in https://github.com/duckdb/duckdb/pull/14514/ but not much point if that library isn't suitable for inclusion ... I just really want to be able to read .csv.bz2 files!

(okay and maybe some others but let's deal with one question at a time)

**Reply by Tishj** [reply]
I don't think that's required, you'd just need a FileSystem extension to support this I believe

**Reply by nickzoic** [reply]
Okay, ummm, I was looking at how the gzip encoding is handled with third_party/miniz in d06566bed67 but if there's an easier way that'd be great too!

I can *read* .csv.bz2 etc in the python library already by passing a file object instead of a filename, but don't seem to be able to write them that way.  That's probably an easier approach.

**Reply by nickzoic** [reply]
Okay I've made a somewhat flakey start on a filesystem extension at https://github.com/nickzoic/duckdb-compression/ and will continue to work on that over the next little while.

**Reply by nickzoic** [reply]
Somewhat related: https://github.com/duckdb/duckdb-python/issues/477 which is my motivation for getting this working :-)

**Reply by nickzoic** [reply]
Ummm, @Tishj could you let me know if https://github.com/nickzoic/duckdb-compression/ is on vaguely the right track, with an eye to eventually contributing it to the community extensions?

**Reply by Tishj** [reply]
Looks like it's on the right track, but I'm hardly the authority on this subject. Registering a FileSystem should make the system pick it based on the read file extension.

**Reply by nickzoic** [reply]
Okay great thanks (the problem with contributing to project core is never knowing if your ideas will get approved, the problem with making plugins is never knowing if they're actually useful or not!)

**Comment by dentiny** [comment]
Hi community , I add a hook in DuckDB core so extensions could register their own compression filesystems; I implemented this [extension](https://github.com/dentiny/duckdb-compression-filesystem) which is supposed to release v2.0
